vde/doc/vdeq.1
.\" Copyright (c) 2004-2005 Renzo Davoli
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139,
.\" USA.

.TH VDEQ 1 "November 18, 2005" "Virtual Distributed Ethernet"
.SH NAME
vdeq \- Virtual Distributed Ethernet wrapper for QEMU virtual machines
.SH SYNOPSIS
.B vdeq 
.B qemu
[
.B \-sock
.I socketpath
[
.I ,socketpath
[
.I ,...
]
]
]
[ qemu_args ]
.br
.B vde\fP{\fIqemu_name\fP} 
[
.B \-sock
.I socketpath
[
.I ,socketpath
[
.I ,...
]
]
]
[ qemu_args ]

.br
.SH DESCRIPTION
\fBvdeq\fP 
is a wrapper to start a QEMU virtual machine connected to a VDE network.
It uses the qemu \fB \-tun-fd \fP parameter to set up the connection with
a vde_switch.

The command
.RS
.br
.B 
vdeq qemu -b c -hda=sampledisk
.RE
starts a qemu machine which boots from the sampledisk image and
has a ne2000 ethernet interface connected to the standard vde_switch.
.br

It is also possible to create symbolic links to the vdeq executable to have
a simpler command. If the link has a name that begins with vde the remaning part
of the name is taken as the qemu command. For example if vdeq is linked to
vdeqemu:
.RS
.br
.B 
vdeqemu -b c -hda=sampledisk
.RE
starts qemu as above.

.RS
.br
.B 
vdeq qemu -sock /tmp/vde.ctl,/tmp/my.ctl -b c -hda=sampledisk
.RE
.RS
.br
.B 
vdeqemu -sock /tmp/vde.ctl,/tmp/my.ctl -b c -hda=sampledisk
.RE
both start qemu with two ethernet interface connected to the standard switch and
to the switch with local socket "/tmp/my.ctl" respectively 

.SH OPTIONS
.TP
.B \-sock 
.I socketpath, 
specify the UNIX socket to be used by local programs for joining the VDE.
The default value is "/tmp/vde.ctl"
It is also possible to indicate several socketpath (comma separated, no blanks):
in this case several interfaces are defined. The first socketpath is connected
to the first ne2k interface and so on.
.br
.TP
.B -h, -help, --help
print a Usage help.
.SH NOTE
Qemu has changed its syntax for networking (cvs Nov.15 2005).
As a temporary solution use vdeoq and vdeoqemu instead of vdeq and vdeqemu 
if you are running a qemu with the old syntax.
.br
.SH SEE ALSO
.BR vde_switch (1),
.BR vde_plug (1),
.BR dpipe (1).
.br
.SH AUTHOR
VDE is a project by Renzo Davoli <renzo@cs.unibo.it>


vde/qemu/vdeq.c
/* Copyright 2003 Renzo Davoli 
 * TNX: 2005.11.18 new syntax mgmt patch by Iain McFarlane <imcfarla@tiscali.co.uk>
 * Licensed under the GPL
 */

#include <config.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <errno.h>
#include <unistd.h>
#include <stdint.h>
#include <libgen.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/uio.h>
#include <sys/poll.h>

#include <vde.h>

#define SWITCH_MAGIC 0xfeedface
#define BUFSIZE 2048
#define ETH_ALEN 6

enum request_type { REQ_NEW_CONTROL };

struct request_v3 {
  uint32_t magic;
  uint32_t version;
  enum request_type type;
  struct sockaddr_un sock;
};


static int send_fd(char *name, int fddata, struct sockaddr_un *datasock, int intno, int group)
{
  int pid = getpid();
  struct request_v3 req;
  int fdctl;

  struct sockaddr_un sock;

  if((fdctl = socket(AF_UNIX, SOCK_STREAM, 0)) < 0){
    perror("socket");
    exit(1);
  }

  sock.sun_family = AF_UNIX;
  snprintf(sock.sun_path, sizeof(sock.sun_path), "%s", name);
  if(connect(fdctl, (struct sockaddr *) &sock, sizeof(sock))){
    perror("connect");
    exit(1);
  }

  req.magic=SWITCH_MAGIC;
  req.version=3;
  req.type=REQ_NEW_CONTROL+((group > 0)?((geteuid()<<8) + group) << 8:0);
  
  req.sock.sun_family=AF_UNIX;
  memset(req.sock.sun_path, 0, sizeof(req.sock.sun_path));
  sprintf(&req.sock.sun_path[1], "%5d-%2d", pid, intno);

  if(bind(fddata, (struct sockaddr *) &req.sock, sizeof(req.sock)) < 0){
    perror("bind");
    exit(1);
  }

  if (send(fdctl,&req,sizeof(req),0) < 0) {
    perror("send");
    exit(1);
  }

  if (recv(fdctl,datasock,sizeof(struct sockaddr_un),0)<0) {
	  perror("recv");
	  exit(1);
  }

  return fdctl;
}

unsigned char bufin[BUFSIZE];

struct pollfd *pollv;

char *filename;
char *vdeqname;
#define NUMW 10

static int countnics(const char *s)
{
	register int nics=1;
	while (*s) {
		if (*s==',') nics++;
		s++;
	}
	return nics;
}

static void usage() 
{
	if (strcmp(vdeqname,"vdeq") != 0 && strncmp(vdeqname,"vde",3)==0) 
		fprintf(stderr,"Usage: %s [-h]\n"
				"\t %s  [-sock sock1 [,sock2...]] qemu_options\n"
				"\t (%s executes a qemu machine named %s)\n", vdeqname,vdeqname,vdeqname,filename);
	else 
		fprintf(stderr,"Usage: %s [-h]\n"
				"\t %s qemu_executable [-sock sock1 [,sock2...]] qemu_options\n", vdeqname, vdeqname);
	exit(0);
}

static void leave()
{
	fprintf(stderr,"qemu exited: %s quits\n", vdeqname);
	exit(0);
}

int main(int argc, char **argv)
{
  int *fddata;
  char *argsock,**sockname;
  struct sockaddr_un *dataout,datain;
  int datainsize;
  int result;
  int group=0;
  int *connected_fd;
  register ssize_t nx;
  int args;
  int newargc;
  char **newargv;
  typedef int pair[2];
  pair *sp;
  register int i;
  int nb_nics;
  int oldsyntax=0;

  vdeqname=basename(argv[0]);
	/* OLD SYNTAX MGMT */
	if (strncmp(vdeqname,"vdeo",4) == 0) {
		oldsyntax=1;
		if (strcmp(vdeqname,"vdeoq") != 0) {
			filename=vdeqname+4;
			args=1;
		}
	}
	else 
		if (strcmp(vdeqname,"vdeq") != 0 && strncmp(vdeqname,"vde",3)==0) {
	  filename=vdeqname+3;
	  args=1;
  }
  else if (argc > 1) {
	  filename=argv[1];
	  args=2;
  } else {
	  usage();
  }
  if ((argc > args && (
			  strcmp(argv[args],"-h")==0 ||
			  strcmp(argv[args],"-help")==0 ||
			  strcmp(argv[args],"--help")==0
			  )) || (
			  strcmp(filename,"-h")==0 ||
			  strcmp(filename,"-help")==0 ||
			  strcmp(filename,"--help")==0
		  )) {
	  usage();
  } else if (argc > args+1 && 
		  (strcmp(argv[args],"-vdesock")==0) ||
		  (strcmp(argv[args],"-sock")==0) ||
		  (strcmp(argv[args],"--sock")==0) ||
		  (strcmp(argv[args],"-unix")==0) ||
		  (strcmp(argv[args],"-s")==0)
	    ){
	  argsock=argv[args+1];
	  args+=2;
  } else
	  //argsock=strdup(VDESTDSOCK);
	  argsock=VDESTDSOCK;

  nb_nics=countnics(argsock);
  if ((sp= (pair *) malloc(nb_nics * 2 * sizeof (int)))<0) {
	  perror("malloc nics");
	  exit(1);
  }

  for (i=0; i<nb_nics; i++) {
  	if (socketpair(AF_UNIX, SOCK_DGRAM, 0, sp[i]) < 0){
	  	perror("socketpair");
	  	exit(1);
	}
  }

  if ((sockname= (char **) malloc(sizeof(char *) * nb_nics))<0) {
	  perror("malloc sockname");
	  exit(1);
  }

  {
	  register char *s=argsock;
	  register char oldch;
	  i=0;
	  do {
		sockname[i++]=s;
		while (*s != ',' && *s != '\0')
			s++;
		oldch=*s;
		*s=0;
		s++;
	  } while (oldch != 0);
  }

/*  printf("-- %s --\n",numfd);
  printf("as %s\n",argsock);
	    for (i=0; i<nb_nics; i++)
		    printf("%d -> %s\n",i,sockname[i]); */
  newargc=argc+1+(2*oldsyntax)+(2*nb_nics)-args;
  if ((newargv=(char **) malloc ((newargc+1)* sizeof(char *))) <0) {
	  perror("malloc");
	  exit(1);
  }

  newargv[0]=filename;
	if (oldsyntax) {
		for (i=0; i<nb_nics; i++) {
			char numfd[10];
			sprintf(numfd,"%d",sp[i][0]);
			newargv[2*i+1]="-tun-fd";
			newargv[2*i+2]=strdup(numfd);
		}
		{
			char nnics[10];
			sprintf(nnics,"%d",nb_nics);
			newargv[2*nb_nics+1]="-nics";
			newargv[2*nb_nics+2]=strdup(nnics);
		}
		for (i=(2*nb_nics)+3;args<argc;i++,args++) newargv[i]=argv[args];
	} else {
		for (i=0; i<nb_nics; i++) {
			char numfd[30];
			sprintf(numfd,"tap,vlan=0,fd=%d",sp[i][0]);
			newargv[2*i+1]="-net";
			newargv[2*i+2]=strdup(numfd);
		}
		for (i=(2*nb_nics)+1;args<argc;i++,args++) newargv[i]=argv[args];
	}

  newargv[i]=0;

  if ((fddata= (int *) malloc(sizeof(int) * nb_nics))<0) {
	  perror("malloc fddata");
	  exit(1);
  }
  if ((connected_fd= (int *) malloc(sizeof(int) * nb_nics))<0) {
	  perror("malloc connected_fd");
	  exit(1);
  }
  if ((dataout= (struct sockaddr_un *) malloc(sizeof(struct sockaddr_un) * nb_nics))<0) {
	  perror("malloc dataout");
	  exit(1);
  }

  if ((pollv= (struct pollfd *) malloc(sizeof(struct pollfd) * 2 * nb_nics))<0) {
	  perror("malloc fddata");
	  exit(1);
  }
  for (i=0; i<nb_nics; i++) {
	  if((fddata[i] = socket(AF_UNIX, SOCK_DGRAM, 0)) < 0){
		  perror("socket");
		  exit(1);
	  }
	  connected_fd[i]=send_fd(sockname[i], fddata[i], &(dataout[i]), i, group);
	  pollv[2*i+1].fd=fddata[i];
	  pollv[2*i].fd=sp[i][1];
	  pollv[2*i].events= pollv[2*i+1].events=POLLIN|POLLHUP;
  }

  if (fork()) {
	  close(0); 
	  signal(SIGCHLD, leave);
	  for (i=0; i<nb_nics; i++) 
		  close(sp[i][0]);
	  for(;;) {
		  result=poll(pollv,2*nb_nics,-1);
		  for (i=0; i<nb_nics; i++) {
			  if (pollv[2*i].revents & POLLHUP || pollv[2*i+1].revents & POLLHUP)
				  break;
			  if (pollv[2*i].revents & POLLIN) {
				  nx=read(sp[i][1],bufin,sizeof(bufin));
				  //fprintf(stderr,"RX from qemu %d\n",nx);
				  //send(connected_fd,bufin,nx,0);
				  sendto(fddata[i],bufin,nx,0,(struct sockaddr *) &(dataout[i]), sizeof(struct sockaddr_un));
			  }
			  if (pollv[2*i+1].revents & POLLIN) {
				  datainsize=sizeof(datain);
				  nx=recvfrom(fddata[i],bufin,BUFSIZE,0,(struct sockaddr *) &datain, &datainsize);
				  //fprintf(stderr,"TX to qemu %d\n",nx);
				  write(sp[i][1],bufin,nx);
			  }
		  }
	  }
  } else {
	  for (i=0; i<nb_nics; i++) {
		  close(sp[i][1]);
		  close(fddata[i]);
		  close(connected_fd[i]);
	  }
	  execvp(filename,newargv);
  }  
  return(0);
}
